// @mistralai/mistralai-gcp/models/components/chatcompletionrequest.js
"use strict";
/*
 * Code generated by Speakeasy (https://speakeasy.com). DO NOT EDIT.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChatCompletionRequest$ = exports.ChatCompletionRequest$outboundSchema = exports.ChatCompletionRequest$inboundSchema = exports.ChatCompletionRequestToolChoice$ = exports.ChatCompletionRequestToolChoice$outboundSchema = exports.ChatCompletionRequestToolChoice$inboundSchema = exports.ChatCompletionRequestMessages$ = exports.ChatCompletionRequestMessages$outboundSchema = exports.ChatCompletionRequestMessages$inboundSchema = exports.ChatCompletionRequestStop$ = exports.ChatCompletionRequestStop$outboundSchema = exports.ChatCompletionRequestStop$inboundSchema = void 0;
exports.chatCompletionRequestStopToJSON = chatCompletionRequestStopToJSON;
exports.chatCompletionRequestStopFromJSON = chatCompletionRequestStopFromJSON;
exports.chatCompletionRequestMessagesToJSON = chatCompletionRequestMessagesToJSON;
exports.chatCompletionRequestMessagesFromJSON = chatCompletionRequestMessagesFromJSON;
exports.chatCompletionRequestToolChoiceToJSON = chatCompletionRequestToolChoiceToJSON;
exports.chatCompletionRequestToolChoiceFromJSON = chatCompletionRequestToolChoiceFromJSON;
exports.chatCompletionRequestToJSON = chatCompletionRequestToJSON;
exports.chatCompletionRequestFromJSON = chatCompletionRequestFromJSON;
const z = __importStar(require("zod"));
const primitives_js_1 = require("../../lib/primitives.js");
const schemas_js_1 = require("../../lib/schemas.js");
const assistantmessage_js_1 = require("./assistantmessage.js");
const prediction_js_1 = require("./prediction.js");
const responseformat_js_1 = require("./responseformat.js");
const systemmessage_js_1 = require("./systemmessage.js");
const tool_js_1 = require("./tool.js");
const toolchoice_js_1 = require("./toolchoice.js");
const toolchoiceenum_js_1 = require("./toolchoiceenum.js");
const toolmessage_js_1 = require("./toolmessage.js");
const usermessage_js_1 = require("./usermessage.js");
/** @internal */
exports.ChatCompletionRequestStop$inboundSchema = z.union([z.string(), z.array(z.string())]);
/** @internal */
exports.ChatCompletionRequestStop$outboundSchema = z.union([z.string(), z.array(z.string())]);
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var ChatCompletionRequestStop$;
(function (ChatCompletionRequestStop$) {
    /** @deprecated use `ChatCompletionRequestStop$inboundSchema` instead. */
    ChatCompletionRequestStop$.inboundSchema = exports.ChatCompletionRequestStop$inboundSchema;
    /** @deprecated use `ChatCompletionRequestStop$outboundSchema` instead. */
    ChatCompletionRequestStop$.outboundSchema = exports.ChatCompletionRequestStop$outboundSchema;
})(ChatCompletionRequestStop$ || (exports.ChatCompletionRequestStop$ = ChatCompletionRequestStop$ = {}));
function chatCompletionRequestStopToJSON(chatCompletionRequestStop) {
    return JSON.stringify(exports.ChatCompletionRequestStop$outboundSchema.parse(chatCompletionRequestStop));
}
function chatCompletionRequestStopFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.ChatCompletionRequestStop$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatCompletionRequestStop' from JSON`);
}
/** @internal */
exports.ChatCompletionRequestMessages$inboundSchema = z.union([
    systemmessage_js_1.SystemMessage$inboundSchema.and(z.object({ role: z.literal("system") }).transform((v) => ({
        role: v.role,
    }))),
    usermessage_js_1.UserMessage$inboundSchema.and(z.object({ role: z.literal("user") }).transform((v) => ({ role: v.role }))),
    assistantmessage_js_1.AssistantMessage$inboundSchema.and(z.object({ role: z.literal("assistant") }).transform((v) => ({
        role: v.role,
    }))),
    toolmessage_js_1.ToolMessage$inboundSchema.and(z.object({ role: z.literal("tool") }).transform((v) => ({ role: v.role }))),
]);
/** @internal */
exports.ChatCompletionRequestMessages$outboundSchema = z.union([
    systemmessage_js_1.SystemMessage$outboundSchema.and(z.object({ role: z.literal("system") }).transform((v) => ({
        role: v.role,
    }))),
    usermessage_js_1.UserMessage$outboundSchema.and(z.object({ role: z.literal("user") }).transform((v) => ({ role: v.role }))),
    assistantmessage_js_1.AssistantMessage$outboundSchema.and(z.object({ role: z.literal("assistant") }).transform((v) => ({
        role: v.role,
    }))),
    toolmessage_js_1.ToolMessage$outboundSchema.and(z.object({ role: z.literal("tool") }).transform((v) => ({ role: v.role }))),
]);
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var ChatCompletionRequestMessages$;
(function (ChatCompletionRequestMessages$) {
    /** @deprecated use `ChatCompletionRequestMessages$inboundSchema` instead. */
    ChatCompletionRequestMessages$.inboundSchema = exports.ChatCompletionRequestMessages$inboundSchema;
    /** @deprecated use `ChatCompletionRequestMessages$outboundSchema` instead. */
    ChatCompletionRequestMessages$.outboundSchema = exports.ChatCompletionRequestMessages$outboundSchema;
})(ChatCompletionRequestMessages$ || (exports.ChatCompletionRequestMessages$ = ChatCompletionRequestMessages$ = {}));
function chatCompletionRequestMessagesToJSON(chatCompletionRequestMessages) {
    return JSON.stringify(exports.ChatCompletionRequestMessages$outboundSchema.parse(chatCompletionRequestMessages));
}
function chatCompletionRequestMessagesFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.ChatCompletionRequestMessages$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatCompletionRequestMessages' from JSON`);
}
/** @internal */
exports.ChatCompletionRequestToolChoice$inboundSchema = z.union([toolchoice_js_1.ToolChoice$inboundSchema, toolchoiceenum_js_1.ToolChoiceEnum$inboundSchema]);
/** @internal */
exports.ChatCompletionRequestToolChoice$outboundSchema = z.union([toolchoice_js_1.ToolChoice$outboundSchema, toolchoiceenum_js_1.ToolChoiceEnum$outboundSchema]);
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var ChatCompletionRequestToolChoice$;
(function (ChatCompletionRequestToolChoice$) {
    /** @deprecated use `ChatCompletionRequestToolChoice$inboundSchema` instead. */
    ChatCompletionRequestToolChoice$.inboundSchema = exports.ChatCompletionRequestToolChoice$inboundSchema;
    /** @deprecated use `ChatCompletionRequestToolChoice$outboundSchema` instead. */
    ChatCompletionRequestToolChoice$.outboundSchema = exports.ChatCompletionRequestToolChoice$outboundSchema;
})(ChatCompletionRequestToolChoice$ || (exports.ChatCompletionRequestToolChoice$ = ChatCompletionRequestToolChoice$ = {}));
function chatCompletionRequestToolChoiceToJSON(chatCompletionRequestToolChoice) {
    return JSON.stringify(exports.ChatCompletionRequestToolChoice$outboundSchema.parse(chatCompletionRequestToolChoice));
}
function chatCompletionRequestToolChoiceFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.ChatCompletionRequestToolChoice$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatCompletionRequestToolChoice' from JSON`);
}
/** @internal */
exports.ChatCompletionRequest$inboundSchema = z.object({
    model: z.string(),
    temperature: z.nullable(z.number()).optional(),
    top_p: z.number().optional(),
    max_tokens: z.nullable(z.number().int()).optional(),
    stream: z.boolean().default(false),
    stop: z.union([z.string(), z.array(z.string())]).optional(),
    random_seed: z.nullable(z.number().int()).optional(),
    messages: z.array(z.union([
        systemmessage_js_1.SystemMessage$inboundSchema.and(z.object({ role: z.literal("system") }).transform((v) => ({
            role: v.role,
        }))),
        usermessage_js_1.UserMessage$inboundSchema.and(z.object({ role: z.literal("user") }).transform((v) => ({
            role: v.role,
        }))),
        assistantmessage_js_1.AssistantMessage$inboundSchema.and(z.object({ role: z.literal("assistant") }).transform((v) => ({
            role: v.role,
        }))),
        toolmessage_js_1.ToolMessage$inboundSchema.and(z.object({ role: z.literal("tool") }).transform((v) => ({
            role: v.role,
        }))),
    ])),
    response_format: responseformat_js_1.ResponseFormat$inboundSchema.optional(),
    tools: z.nullable(z.array(tool_js_1.Tool$inboundSchema)).optional(),
    tool_choice: z.union([toolchoice_js_1.ToolChoice$inboundSchema, toolchoiceenum_js_1.ToolChoiceEnum$inboundSchema])
        .optional(),
    presence_penalty: z.number().optional(),
    frequency_penalty: z.number().optional(),
    n: z.nullable(z.number().int()).optional(),
    prediction: prediction_js_1.Prediction$inboundSchema.optional(),
    parallel_tool_calls: z.boolean().optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        "top_p": "topP",
        "max_tokens": "maxTokens",
        "random_seed": "randomSeed",
        "response_format": "responseFormat",
        "tool_choice": "toolChoice",
        "presence_penalty": "presencePenalty",
        "frequency_penalty": "frequencyPenalty",
        "parallel_tool_calls": "parallelToolCalls",
    });
});
/** @internal */
exports.ChatCompletionRequest$outboundSchema = z.object({
    model: z.string(),
    temperature: z.nullable(z.number()).optional(),
    topP: z.number().optional(),
    maxTokens: z.nullable(z.number().int()).optional(),
    stream: z.boolean().default(false),
    stop: z.union([z.string(), z.array(z.string())]).optional(),
    randomSeed: z.nullable(z.number().int()).optional(),
    messages: z.array(z.union([
        systemmessage_js_1.SystemMessage$outboundSchema.and(z.object({ role: z.literal("system") }).transform((v) => ({
            role: v.role,
        }))),
        usermessage_js_1.UserMessage$outboundSchema.and(z.object({ role: z.literal("user") }).transform((v) => ({
            role: v.role,
        }))),
        assistantmessage_js_1.AssistantMessage$outboundSchema.and(z.object({ role: z.literal("assistant") }).transform((v) => ({
            role: v.role,
        }))),
        toolmessage_js_1.ToolMessage$outboundSchema.and(z.object({ role: z.literal("tool") }).transform((v) => ({
            role: v.role,
        }))),
    ])),
    responseFormat: responseformat_js_1.ResponseFormat$outboundSchema.optional(),
    tools: z.nullable(z.array(tool_js_1.Tool$outboundSchema)).optional(),
    toolChoice: z.union([
        toolchoice_js_1.ToolChoice$outboundSchema,
        toolchoiceenum_js_1.ToolChoiceEnum$outboundSchema,
    ]).optional(),
    presencePenalty: z.number().optional(),
    frequencyPenalty: z.number().optional(),
    n: z.nullable(z.number().int()).optional(),
    prediction: prediction_js_1.Prediction$outboundSchema.optional(),
    parallelToolCalls: z.boolean().optional(),
}).transform((v) => {
    return (0, primitives_js_1.remap)(v, {
        topP: "top_p",
        maxTokens: "max_tokens",
        randomSeed: "random_seed",
        responseFormat: "response_format",
        toolChoice: "tool_choice",
        presencePenalty: "presence_penalty",
        frequencyPenalty: "frequency_penalty",
        parallelToolCalls: "parallel_tool_calls",
    });
});
/**
 * @internal
 * @deprecated This namespace will be removed in future versions. Use schemas and types that are exported directly from this module.
 */
var ChatCompletionRequest$;
(function (ChatCompletionRequest$) {
    /** @deprecated use `ChatCompletionRequest$inboundSchema` instead. */
    ChatCompletionRequest$.inboundSchema = exports.ChatCompletionRequest$inboundSchema;
    /** @deprecated use `ChatCompletionRequest$outboundSchema` instead. */
    ChatCompletionRequest$.outboundSchema = exports.ChatCompletionRequest$outboundSchema;
})(ChatCompletionRequest$ || (exports.ChatCompletionRequest$ = ChatCompletionRequest$ = {}));
function chatCompletionRequestToJSON(chatCompletionRequest) {
    return JSON.stringify(exports.ChatCompletionRequest$outboundSchema.parse(chatCompletionRequest));
}
function chatCompletionRequestFromJSON(jsonString) {
    return (0, schemas_js_1.safeParse)(jsonString, (x) => exports.ChatCompletionRequest$inboundSchema.parse(JSON.parse(x)), `Failed to parse 'ChatCompletionRequest' from JSON`);
}
